### ISSUE TITLE
Problem with pub keyword in macros

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#18317]
Hey,

I have an issue on my macros use. The source code of my macro is [here](https://gist.github.com/GuillaumeGomez/8f78b43c19394e2213a6). I can't put the pub keyword and if I call my macro like this :

```Rust
pub impl_GObjectFunctions!(t1, t2)
```

The methods aren't public. Am I doing something wrong or is it a compiler's bug ?

Thanks in advance.

**Comment by hugoduncan** [comment]
I just encountered this.  I can reproduce it with:

```rust
macro_rules! defn {
    ($n:ident) => (
        fn $n (&self) -> i32 {
            println!("{}", stringify!($n));
            1
        }
        )
}

#[derive(Copy)]
pub struct S;

impl S {
    // Attempt to define a public method via a macro.
    // The resulting method ends up being non-public.
    pub defn!(f);
}
```

**Comment by GuillaumeGomez** [comment]
Now let's just hope it'll be fixed sooner or later !

**Comment by kmcallister** [comment]
Hm, what should `pub item_macro!();` actually mean?  A macro in item position can expand to zero or more items.  Should the `pub` qualifier apply to all of them, or should it be an error to use it with a multi-item macro?

We might want to forbid `pub` in this position for 1.0, and come back to the issue later.  I had a design for [item macro sugar](https://github.com/rust-lang/rfcs/pull/453#issuecomment-62788265) that I think would be rather nice.

The current workaround is to write

```rust
item_macro!(pub thing ...);
```

The precise syntax is up to the macro author and can accommodate multiple items.

This workaround would be less annoying if we had a "0 or 1" repeat operator in macros.

**Comment by hugoduncan** [comment]
@kmcallister Could you be a little more explicit about the workaround - I can't seem to get it to work.

**Comment by kmcallister** [comment]
@hugoduncan: [Here's an example](https://github.com/rust-lang/rust/blob/f9aeea7cb7865a2b82e7102837daabbe549177ea/src/librustc/lint/mod.rs#L85-L97) from the compiler itself.

[This approach from lazy-static](https://github.com/Kimundi/lazy-static.rs/blob/b48b0c551087af9d598a0452f4e3973d98d4419b/src/lazy_static.rs#L75-L80) looks better, with less duplicated code.

**Comment by hugoduncan** [comment]
@kmcallister  Thanks for the examples.  If I understand them correctly (neither of which generate `fn` forms, iiuc), I should be matching on `pub` and inserting a literal `pub` in the expansion, but [this](http://is.gd/gEHs3v) does not seem to work :

```rust
macro_rules! defn {
    (pub $n:ident) => (
        pub fn $n (&self) -> i32 { 1 } )
}

#[derive(Copy)]
pub struct S;

impl S {
    defn!(pub f);
}


#[cfg(test)]
mod tests {
    use super::S;

    #[test]
    fn test_f() {
        let s=S;
        assert_eq!(1, s.f());
    }
}
```

**Comment by kmcallister** [comment]
It works for me on rustc 1.0.0-dev (e4e7aa285 2015-02-16).  The playpen is just [too old](http://is.gd/aaoGCp) to have my fix for #17436.

**Comment by hugoduncan** [comment]
Thanks for the confirmation.  I'll be looking forward to a new nightly for once :)
